File: 3665.cpp
#include <iostream>
#include <queue>
#include <vector>

#include "modules/FastIO.h"
#include "modules/Graph.h"
#include "modules/GraphAlgs.h"
#include "modules/Input.h"
#include "modules/Output.h"
#include "modules/Types.h"

void loop() {
	int n;
	std::cin >> n;
	std::vector<int> teams = Input::inlineToVec<int>(n);
	for (int& team: teams) team --;

	MatrixGraph<None, int> graph(n);

	for (int winner = 0; winner < n; winner++) {
		for (int loser = winner + 1; loser < n; loser++) {
			graph.edge(teams[winner], teams[loser]) = true;
		}
	}

	int m;
	std::cin >> m;
	for (int i = 0; i < m; i++) {
		int a, b;
		std::cin >> a >> b;
		a --;
		b --;

		graph.edge(a, b) = !graph.edge(a, b);
		graph.edge(b, a) = !graph.edge(b, a);
	}

	std::vector<int> dependencies = GraphAlgs::dependencies(graph);

	std::queue<int> queue;

	bool uncertain = false;

	for (int i = 0; i < n; i++) {
		if (dependencies[i] == 0) {
			queue.push(i);
		}
	}

	std::vector<int> result;
	result.reserve(n);

	for (int i = 0; i < n; i++) {
		if (queue.empty()) {
			std::cout << "IMPOSSIBLE\n";
			return;
		}

		if (queue.size() > 1) {
			uncertain = true;
		}

		int parent = queue.front();
		queue.pop();
		result.push_back(parent + 1);

		graph.forEachChild(parent, [&queue, &dependencies, &result](int child, None value) {
				dependencies[child] --;
				if (dependencies[child] == 0) {
					queue.push(child);
				}
				});
	}
	if (uncertain)
		std::cout << "?\n";
	else
		std::cout << result << '\n';
}

int main() {
	FastIO();
	int t;
	std::cin >> t;
	for (int i = 0; i < t; i++) {
		loop();
	}
}

File: solutions/3665.py
from collections import deque

def main():
    n = int(input())

    matrix = [[False] * n for _ in range(n)]

    positions = list(map(int, input().split()))

    for i in range(n):
        for j in range(i + 1, n):
            matrix[positions[i] - 1][positions[j] - 1] = True

    m = int(input())

    for _ in range(m):
        a, b = tuple(map(int, input().split()))
        a -= 1
        b -= 1

        if matrix[a][b] == False:
            matrix[a][b] = True
            matrix[b][a] = False
        else:
            matrix[a][b] = False
            matrix[b][a] = True

    dependencies = [0] * n

    for parent in range(n):
        for child in range(n):
            if matrix[parent][child]:
                dependencies[child] += 1

    queue = deque()

    for i in range(n):
        if dependencies[i] == 0:
            queue.append(i)

    maybe = False

    result = []

    for _ in range(n):
        if len(queue) == 0:
            print('IMPOSSIBLE')
            return
        if len(queue) >= 2:
            maybe = True

        parent = queue.pop()

        result.append(parent + 1)

        for child in range(n):
            if not matrix[parent][child]: continue

            dependencies[child] -= 1

            if dependencies[child] == 0:
                queue.append(child)

    if maybe:
        print('?')
    else:
        print(*result)

for _ in range(int(input())):
    main()
